src/engines-experimental/stree/persistent_b_tree.h

```diff
@@ -1750,19 +1932,445 @@ b_tree_base<TKey, TValue, degree>::insert_descend(pool_base &pop, const_referenc
 	;
 }
 
-} // namespace internal
+template <typename Key, typename T, typename Compare, size_t degree>
+typename b_tree_base<Key, T, Compare, degree>::inner_type::const_iterator
+b_tree_base<Key, T, Compare, degree>::split_half(pool_base &pop,
+						 persistent_ptr<inner_type> &node,
+						 persistent_ptr<node_t> &left,
+						 persistent_ptr<node_t> &right)
+{
+	typename inner_type::const_iterator middle = node->begin() + node->size() / 2;
+	typename inner_type::const_iterator l_begin = node->begin();
+	typename inner_type::const_iterator l_end = middle;
+	typename inner_type::const_iterator r_begin = middle + 1;
+	typename inner_type::const_iterator r_end = node->end();
+	pmem::obj::transaction::run(pop, [&] {
+		cast_inner(left) =
+			allocate_inner(pop, node->level(), l_begin, l_end, node.get());
+		cast_inner(right) =
+			allocate_inner(pop, node->level(), r_begin, r_end, node.get());
+	});
+
+	return middle;
+}
+
+template <typename Key, typename T, typename Compare, size_t degree>
+void b_tree_base<Key, T, Compare, degree>::split_inner_node(pool_base &pop,
+							    inner_node_pptr &src_node,
+							    inner_type *parent_node,
+							    node_pptr &left,
+							    node_pptr &right)
+{
+	typename inner_type::const_iterator partition_point =
+		split_half(pop, src_node, left, right);
+	pmem::obj::transaction::run(pop, [&] {
+		if (parent_node) {
+			parent_node->update_splitted_child(pop, *partition_point, left,
+							   right);
+		} else {
+			assert(root == src_node);
+			create_new_root(pop, *partition_point, left, right);
+		}
+	});
+}
+
+template <typename Key, typename T, typename Compare, size_t degree>
+template <typename K, typename M>
+typename b_tree_base<Key, T, Compare, degree>::iterator
+b_tree_base<Key, T, Compare, degree>::split_leaf_node(pool_base &pop,
+						      inner_type *parent_node,
+						      leaf_node_pptr &split_leaf, K &&key,
+						      M &&obj)
+{
+	assert(split_leaf->full());
+
+	leaf_node_pptr node;
+	typename leaf_type::iterator entry_it;
+	// move second half into node and insert new element where needed
+	pmem::obj::transaction::run(pop, [&] {
+		node = allocate_leaf(pop);
+		entry_it = node->move(pop, split_leaf, std::forward<K>(key),
+				      std::forward<M>(obj));
+		// re-set node's pointers
+		node->set_next(split_leaf->get_next());
+		node->set_prev(split_leaf);
+		if (split_leaf->get_next()) {
+			split_leaf->get_next()->set_prev(node);
+		}
+		split_leaf->set_next(node);
+		// take care of parent node
+		if (parent_node) {
+			parent_node->update_splitted_child(
+				pop, split_leaf->back().first,
+				reinterpret_cast<node_pptr &>(split_leaf),
+				reinterpret_cast<node_pptr &>(node));
+		} else {
+			cast_inner(root) = allocate_inner(pop, root->level() + 1,
+							  split_leaf->back().first,
+							  split_leaf, node);
+		}
+	});
+
+	assert(entry_it != node->end());
+	assert(entry_it->first == key);
+	assert(entry_it->second == obj);
```

Comment:
== operator should be defined in case a class is used as template, very possible for value. Should document this for ease.